level00/PropertyTests.hs

```diff
@@ -112,7 +113,45 @@ genTree genKV = fromList <$> Gen.list (Range.linear 0 100) genKV
 genMyBTreeVal :: MonadGen m => m (Int, Char)
 genMyBTreeVal = liftA2 (,) (Gen.int (Range.linear (-100) 100)) (Gen.enum 'a' 'z')
 
--- We're not ready to write a property test for inserting values into our binary
+-- We can also use property based testing to ensure that when we
+-- implement a typeclass instance, that implementation will comply
+-- with the laws of that typeclass. This is important as often the
+-- laws of a typeclass cannot be expressed in the types.
+--
+-- NOTE: There are _two_ valid 'Eq' instances we could write:
+-- - Structural equality
+-- - Matching set of (key, value) pairs.
+--
+-- Implement the 'Eq' instance for 'MyBTree' that compares for two
+-- trees having a matching set of (key, value) pairs.
+--
+instance (Eq k, Eq a) => Eq (MyBTree k a) where
+  (==) :: MyBTree k a -> MyBTree k a -> Bool
+  (==) a b = toListWithKey a == toListWithKey b
+
+prop_MyBTree_LawfulEqInstance :: Property
+prop_MyBTree_LawfulEqInstance = property $ do
+  (x,y,z) <- forAll $ (,,)
+             <$> genTree genMyBTreeVal
+             <*> genTree genMyBTreeVal
+             <*> genTree genMyBTreeVal
+
+  annotate "Reflexivity: x == x = True"
+  (x == x) === True
+
+  annotate "Symmetry: x == y = y == x"
+  (x == y) === (y == x)
+
+  annotate "Transitivity: if x == y && y == x = True then x == z = True"
```

Comment:
You could probably drop the `= True` from these?

Reply:
They're copied verbatim from the `Eq` docs, which I haven't included a link to. I will expand to include a link.